File: kotlin/src/main/kotlin/15.X3sum.kt
import org.junit.Assert.assertEquals
import org.junit.Test
// https://leetcode.com/problems/3sum


class X3sum {
  fun threeSum(nums: IntArray): List<List<Int>> {
    val n = nums.size
    nums.sort()     // to help uniqueness
    val lastIndex = hashMapOf<Int, Int>()
    for((i, v) in nums.withIndex())
      lastIndex[v] = i

    val res = mutableListOf<List<Int>>()
    for(i in 0 until n) {
      if(i-1 >= 0 && nums[i] == nums[i-1])
        continue
      for (j in i + 1 until n) {
        if(j-1 > i && nums[j] == nums[j-1])
          continue
        val s = nums[i] + nums[j]
        if(j < lastIndex.getOrDefault(-s, 0))
          res.add(listOf(nums[i], nums[j], -s))
      }
    }
    return res
  }

  @Test
  fun test1(){
    assertEquals(listOf(listOf(-1, -1, 2), listOf(-1, 0, 1)), threeSum(intArrayOf(-1, 0, 1, 2, -1, -4)))
  }

  @Test
  fun test2(){
    assertEquals(listOf(listOf(0, 0, 0)), threeSum(intArrayOf(0, 0, 0, 0)))
  }

}

File: python/15_X3sum.py
from unittest import TestCase
# https://leetcode.com/problems/3sum


class X3sum(TestCase):
    def threeSum(self, nums: 'List[int]') -> 'List[List[int]]':
        nums.sort()
        n, res = len(nums), []
        for i in range(n):
            if i-1 >= 0 and nums[i] == nums[i-1]: continue
            j, k = i+1, n-1
            while j < k:
                a = [nums[i], nums[j], nums[k]]
                s = sum(a)
                if s == 0:
                    res.append(a)
                    while j+1 < k and nums[j+1] == nums[j]:
                        j += 1
                    while k-1 > j and nums[k-1] == nums[k]:
                        k -= 1
                if s <= 0: j += 1
                if s >= 0: k -= 1
        return res

    def test1(self):
        self.assertEqual([[-1, -1, 2], [-1, 0, 1]], self.threeSum([-1, 0, 1, 2, -1, -4]))

